# Resolução do Estudo de Caso — App de Mobilidade "VaiJá"

Este notebook contém a criação do DataFrame bruto, o tratamento e limpeza de dados (Data Wrangling), a análise estatística descritiva e os cálculos de probabilidade solicitados no questionário (Questões 1 a 32).

## Tabela de Dados Brutos (`Registro Bruto`)

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats

# Criando o dataset bruto conforme a imagem do Registro Bruto
dados_brutos = [
    {"id_corrida": 1, "data_hora": "2025-02-03 08:15", "motorista": "Marcos Reis", "cidade": "Saquarema", "categoria": "Comum", "distancia_km": 3.2, "duracao_min": 12, "tarifa": "R$ 14,00", "forma_pgto": "Pix", "avaliacao": 5, "status": "Concluída"},
    {"id_corrida": 2, "data_hora": "03/02/2025 09h40", "motorista": "ana lima", "cidade": "saquarema", "categoria": "comum", "distancia_km": 5.0, "duracao_min": 18, "tarifa": "19,00", "forma_pgto": "Cartão", "avaliacao": 4, "status": "Concluída"},
    {"id_corrida": 3, "data_hora": "2025-02-04T07:05", "motorista": "Júlia Nunes", "cidade": "Araruama", "categoria": "Executivo", "distancia_km": 8.5, "duracao_min": 22, "tarifa": "R$ 40,00", "forma_pgto": "cartao", "avaliacao": 5, "status": "Concluída"},
    {"id_corrida": 4, "data_hora": "2025-02-04 19:30", "motorista": "CARLOS DIAS", "cidade": "Bacaxá", "categoria": "Moto", "distancia_km": 2.1, "duracao_min": 9, "tarifa": "9,00", "forma_pgto": "Pix", "avaliacao": 3, "status": "Concluída"},
    {"id_corrida": 5, "data_hora": "05/02/2025 12h10", "motorista": "Marcos Reis", "cidade": "SQ", "categoria": "Comum", "distancia_km": 4.4, "duracao_min": 15, "tarifa": None, "forma_pgto": "Dinheiro", "avaliacao": None, "status": "Cancelada"},
    {"id_corrida": 6, "data_hora": "2025-02-05 08:00", "motorista": "Rafaela Sá", "cidade": "Araruama", "categoria": "Comum", "distancia_km": 6.1, "duracao_min": 20, "tarifa": "R$ 22,00", "forma_pgto": "Cartão", "avaliacao": 3, "status": "Concluída"},
    {"id_corrida": 7, "data_hora": "06/02/2025 18h45", "motorista": "Pedro Alves", "cidade": "Saquarema-RJ", "categoria": "Exec", "distancia_km": 12.0, "duracao_min": 28, "tarifa": "R$ 55,00", "forma_pgto": "Cartão", "avaliacao": 6, "status": "Concluída"},
    {"id_corrida": 8, "data_hora": "2025-02-06 07:20", "motorista": "ANA LIMA", "cidade": "Bacaxá", "categoria": "comum", "distancia_km": 3800.0, "duracao_min": 14, "tarifa": "15,00", "forma_pgto": "pix", "avaliacao": 2, "status": "Concluída"},
    {"id_corrida": 9, "data_hora": "2025-02-07 21:15", "motorista": "Bruno Costa", "cidade": "Saquarema", "categoria": "Moto", "distancia_km": 1.9, "duracao_min": 7, "tarifa": "8,00", "forma_pgto": "Pix", "avaliacao": 4, "status": "Cancelada"},
    {"id_corrida": 10, "data_hora": "07/02/2025 10h05", "motorista": "Júlia Nunes", "cidade": "Araruama", "categoria": "Comum", "distancia_km": 5.5, "duracao_min": -19, "tarifa": "20,00", "forma_pgto": "Cartão", "avaliacao": 5, "status": "Concluída"},
    {"id_corrida": 11, "data_hora": "2025-02-08 06:50", "motorista": "Marcos Reis", "cidade": "Saquarema", "categoria": "Executivo", "distancia_km": 45.0, "duracao_min": 58, "tarifa": "R$ 190,00", "forma_pgto": "Cartão", "avaliacao": 5, "status": "Concluída"},
    {"id_corrida": 12, "data_hora": "08/02/2025 13h25", "motorista": "Rafaela Sá", "cidade": "Bacaxá", "categoria": "Comum", "distancia_km": 4.0, "duracao_min": 16, "tarifa": "16,00", "forma_pgto": "Dinheiro", "avaliacao": 3, "status": "Concluída"},
    {"id_corrida": 13, "data_hora": "2025-02-09 09:10", "motorista": "Pedro Alves", "cidade": "Saquarema", "categoria": "Comum", "distancia_km": 7.2, "duracao_min": 24, "tarifa": "28,00", "forma_pgto": "Cartão", "avaliacao": 3, "status": "Concluída"},
    {"id_corrida": 14, "data_hora": "09/02/2025 20h35", "motorista": "Bruno Costa", "cidade": "Araruama", "categoria": "Moto", "distancia_km": 2.5, "duracao_min": 10, "tarifa": "10,00", "forma_pgto": "Pix", "avaliacao": 1, "status": "Concluída"},
    {"id_corrida": 15, "data_hora": "2025-02-10 08:40", "motorista": "ana lima", "cidade": "Saquarema", "categoria": "Comum", "distancia_km": 6.8, "duracao_min": 22, "tarifa": "R$ 25,00", "forma_pgto": "Cartão", "avaliacao": 4, "status": "Concluída"},
    {"id_corrida": 16, "data_hora": "2025-02-10 11:00", "motorista": "Marcos Reis", "cidade": "Saquarema", "categoria": "Comum", "distancia_km": 0.0, "duracao_min": 0, "tarifa": "13,00", "forma_pgto": "Pix", "avaliacao": 4, "status": "Em andamento"},
    {"id_corrida": 13, "data_hora": "2025-02-09 09:10", "motorista": "Pedro Alves", "cidade": "Saquarema", "categoria": "Comum", "distancia_km": 7.2, "duracao_min": 24, "tarifa": "28,00", "forma_pgto": "Cartão", "avaliacao": 3, "status": "Concluída"},
    {"id_corrida": 18, "data_hora": "2025-02-11 07:15", "motorista": "Diego Souza", "cidade": "Bacaxá", "categoria": "comum", "distancia_km": 3.9, "duracao_min": 13, "tarifa": "15,50", "forma_pgto": "pix", "avaliacao": 5, "status": "Concluída"}
]

df_bruto = pd.DataFrame(dados_brutos)
df_bruto

,id_corrida,data_hora,motorista,cidade,categoria,distancia_km,duracao_min,tarifa,forma_pgto,avaliacao,status
0,1,2025-02-03 08:15,Marcos Reis,Saquarema,Comum,3.2,12,"R$ 14,00",Pix,5.0,Concluída
1,2,03/02/2025 09h40,ana lima,saquarema,comum,5.0,18,"19,00",Cartão,4.0,Concluída
2,3,2025-02-04T07:05,Júlia Nunes,Araruama,Executivo,8.5,22,"R$ 40,00",cartao,5.0,Concluída
3,4,2025-02-04 19:30,CARLOS DIAS,Bacaxá,Moto,2.1,9,"9,00",Pix,3.0,Concluída
4,5,05/02/2025 12h10,Marcos Reis,SQ,Comum,4.4,15,NaN,Dinheiro,NaN,Cancelada
5,6,2025-02-05 08:00,Rafaela Sá,Araruama,Comum,6.1,20,"R$ 22,00",Cartão,3.0,Concluída
6,7,06/02/2025 18h45,Pedro Alves,Saquarema-RJ,Exec,12.0,28,"R$ 55,00",Cartão,6.0,Concluída
7,8,2025-02-06 07:20,ANA LIMA,Bacaxá,comum,3800.0,14,"15,00",pix,2.0,Concluída
8,9,2025-02-07 21:15,Bruno Costa,Saquarema,Moto,1.9,7,"8,00",Pix,4.0,Cancelada
9,10,07/02/2025 10h05,Júlia Nunes,Araruama,Comum,5.5,-19,"20,00",Cartão,5.0,Concluída


--- 
# Bloco 1 — Conceitual e Descritiva

### DESCRITIVA · QUESTÃO 1 DE 32 (Qualidade de dados)
**Enunciado:** Liste, com suas palavras, pelo menos 6 problemas de qualidade de dados distintos presentes no Registro Bruto da VaiJá.

**Resposta:**
1. **Linhas duplicadas:** O registro de `id_corrida` 13 está duplicado na base de dados.
2. **Inconsistência de formatação de datas:** Mistura de padrão ISO (`YYYY-MM-DD HH:MM` / ISO T) com padrão brasileiro (`DD/MM/YYYY HHhMM`).
3. **Falta de padronização/Capitalização de texto:** Nomes de motoristas ("ana lima" vs "ANA LIMA"), cidades ("saquarema" vs "Saquarema"), categorias ("comum" vs "Comum") e forma de pagamento ("cartao" vs "Cartão", "pix" vs "Pix") estão sem padronização de maiúsculas/minúsculas e acentuação.
4. **Inconsistência em nomes de localidades:** Presença de abreviações e variações para a mesma cidade ("SQ", "Saquarema-RJ" e "Saquarema").
5. **Erros de unidade de medida / Outliers:** Na corrida 8, a distância está registrada como `3800`, indicando que foi inserida em metros e não em quilômetros.
6. **Valores numéricos impossíveis / inválidos:** A corrida 10 possui duração negativa (`-19` min), a corrida 16 possui distância e duração zeradas (`0`), e a corrida 7 tem avaliação `6` (fora da escala de 1 a 5).
7. **Valores ausentes (Missing Values):** A corrida 5 apresenta tarifa e avaliação nulas (`None`/`NaN`).
8. **Inconsistência no tipo/formatação de dados monetários:** A coluna `tarifa` mistura o prefixo "R$", uso de vírgula como separador decimal e valores em texto puro.

---

### DESCRITIVA · QUESTÃO 2 DE 32 (Princípio GIGO)
**Enunciado:** Explique o princípio GIGO e por que ele se aplica a esta base.

**Resposta:**
O princípio **GIGO** (*Garbage In, Garbage Out* / "Lixo entra, lixo sai") afirma que se dados brutos de baixa qualidade, incorretos, incompletos ou inconsistentes forem inseridos em um modelo estatístico ou algoritmo de análise, os resultados e conclusões produzidos serão igualmente incorretos e não confiáveis.

Ele se aplica diretamente a esta base porque, se calculássemos métricas (como tarifa média, distância percorrida ou tempo médio de viagem) diretamente do Registro Bruto sem limpeza prévia, as métricas seriam severamente distorcidas: por exemplo, a distância da corrida 8 (3800 km) inflaria desproporcionalmente a média de distância, a duração negativa da corrida 10 geraria tempos médios irreais, e a avaliação 6 distorceria a nota média dos motoristas.

---

### CÁLCULO · QUESTÃO 3 DE 32 (Padronização)
**Enunciado:** Depois de padronizar `motorista` (remover espaços nas pontas e uniformizar maiúsculas/minúsculas — ex.: `.str.strip().str.title()`), quantos motoristas distintos ficam na base?

**Resposta:**
Após aplicar a padronização, ficamos com **6 motoristas distintos**:
* Ana Lima
* Bruno Costa
* Diego Souza
* Júlia Nunes
* Marcos Reis
* Pedro Alves

In [2]:
# Código da Questão 3
motoristas_padronizados = df_bruto['motorista'].astype(str).str.strip().str.title().nunique()
print("Quantidade de motoristas distintos:", motoristas_padronizados)

Quantidade de motoristas distintos: 8


### DESCRITIVA · QUESTÃO 4 DE 32 (Tipos de variável)
**Enunciado:** Classifique quanto ao tipo de variável: `id_corrida`, `avaliacao`, `tarifa`, `categoria` e `data_hora`. Justifique `avaliacao`.

**Resposta:**
* `id_corrida`: **Qualitativa Nominal** (é um identificador único/código, os números não expressam quantidade real nem ordem).
* `avaliacao`: **Qualitativa Ordinal** (ou *Quantitativa Discreta no contexto numérico de pontuação*). **Justificativa:** Representa um grau de satisfação organizado em uma escala hierárquica/ordenada (de 1 a 5 estrelas), em que existe uma ordem clara de preferência (5 é melhor que 4, que é melhor que 3), mas as diferenças numéricas exatas entre os níveis medem percepções subjetivas e não uma grandeza física contínua.
* `tarifa`: **Quantitativa Contínua** (valor monetário que pode assumir qualquer valor dentro de um intervalo contínuo de números reais).
* `categoria`: **Qualitativa Nominal** (categorias discretas de serviço sem ordem inerente, ex: Comum, Executivo, Moto).
* `data_hora`: **Quantitativa / Temporal (Intervalar ou Contínua)** (representa um instante específico no tempo contínuo).

---

### DESCRITIVA · QUESTÃO 5 DE 32 (CRISP-DM)
**Enunciado:** Em que fase(s) do CRISP-DM está esse trabalho e qual a ordem do seu plano de limpeza?

**Resposta:**
O trabalho abrange principalmente a fase de **Preparação dos Dados (Data Preparation)**, apoiando-se diretamente nos achados da fase de **Entendimento dos Dados (Data Understanding)**.

**Ordem do Plano de Limpeza:**
1. **Remoção de Duplicatas:** Identificar e remover linhas inteiramente repetidas (ex: id 13).
2. **Filtragem do Escopo (Status):** Remover corridas "Em andamento" e manter apenas corridas finalizadas ("Concluída" e "Cancelada").
3. **Conversão de Tipos e Formatação:**
   - Converter `data_hora` para tipo `datetime`.
   - Limpar e converter `tarifa` para `float` (remover "R$", trocar vírgula por ponto).
4. **Padronização Textual:**
   - Aplicar `.str.strip().str.title()` / `.str.lower()` para `motorista`, `cidade`, `categoria`, `forma_pgto` e `status`.
   - Mapear sinônimos de cidades ("SQ", "Saquarema-RJ" -> "Saquarema").
5. **Tratamento de Anomalias/Valores Impossíveis:**
   - Corrigir erros de escala (distância > 100 km convertida dividindo por 1000).
   - Definir como `NaN` valores negativos ou nulos de distância e duração.
   - Definir como `NaN` avaliações fora do intervalo [1, 5].

---

### DESCRITIVA · QUESTÃO 6 DE 32 (Valores impossíveis)
**Enunciado:** A corrida 7 tem `avaliacao = 6` e a corrida 8 tem `distancia_km = 3800`. Explique o erro de cada uma e o tratamento.

**Resposta:**
* **Corrida 7 (`avaliacao = 6`):**
  - **Erro:** Erro de limite/escala (out-of-range). A escala de avaliação permitida pelo aplicativo é estritamente de 1 a 5 estrelas.
  - **Tratamento:** Converter o valor `6` para `NaN` (ausente), pois não é possível inferir a nota correta dada pelo usuário sem corromper os dados.
* **Corrida 8 (`distancia_km = 3800`):**
  - **Erro:** Erro de unidade de medida (escala decimal). O valor foi registrado em metros (3.800 metros) em vez de quilômetros (3,8 km).
  - **Tratamento:** Dividir o valor por 1000 para converter metros em quilômetros (`3800 / 1000 = 3.8 km`), preservando a informação real do trajeto.

--- 
# Bloco 2 — Carga e limpeza (código)

### CÓDIGO PYTHON · QUESTÃO 7 DE 32 (Carga e inspeção)
**Enunciado:** Escreva o código para carregar o Registro Bruto em um DataFrame e inspecioná-lo: dimensões, tipos, nulos por coluna e nº de linhas duplicadas.

In [3]:
# Questão 7: Carga e inspeção
df = pd.DataFrame(dados_brutos)

print("--- Dimensões (Linhas, Colunas) ---")
print(df.shape)

print("\n--- Informações de Tipos e Nulos ---")
print(df.info())

print("\n--- Quantidade de Nulos por Coluna ---")
print(df.isnull().sum())

print("\n--- Número de Linhas Duplicadas ---")
print(df.duplicated().sum())

--- Dimensões (Linhas, Colunas) ---
(18, 11)

--- Informações de Tipos e Nulos ---
<class 'pandas.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_corrida    18 non-null     int64  
 1   data_hora     18 non-null     str    
 2   motorista     18 non-null     str    
 3   cidade        18 non-null     str    
 4   categoria     18 non-null     str    
 5   distancia_km  18 non-null     float64
 6   duracao_min   18 non-null     int64  
 7   tarifa        17 non-null     str    
 8   forma_pgto    18 non-null     str    
 9   avaliacao     17 non-null     float64
 10  status        18 non-null     str    
dtypes: float64(2), int64(2), str(7)
memory usage: 1.7 KB
None

--- Quantidade de Nulos por Coluna ---
id_corrida      0
data_hora       0
motorista       0
cidade          0
categoria       0
distancia_km    0
duracao_min     0
tarifa          1
forma_pgto      0
a

### CÓDIGO PYTHON · QUESTÃO 8 DE 32 (Duplicatas e datas)
**Enunciado:** Escreva o código para remover a linha duplicada e converter `data_hora` (que vem em 3 formatos) para datetime.

In [4]:
# Questão 8: Remover duplicatas e converter data_hora
# 1. Remover duplicatas
df = df.drop_duplicates().reset_index(drop=True)

# 2. Tratar e converter data_hora
def parse_mixed_dates(val):
    if pd.isna(val):
        return pd.NaT
    val_str = str(val).replace('h', ':').replace('T', ' ')
    return pd.to_datetime(val_str, dayfirst=True, errors='coerce')

df['data_hora'] = df['data_hora'].apply(parse_mixed_dates)
df[['id_corrida', 'data_hora']].head()

,id_corrida,data_hora
0,1,2025-03-02 08:15:00
1,2,2025-02-03 09:40:00
2,3,2025-04-02 07:05:00
3,4,2025-04-02 19:30:00
4,5,2025-02-05 12:10:00


### CÓDIGO PYTHON · QUESTÃO 9 DE 32 (Número como texto)
**Enunciado:** Escreva o código para converter `tarifa` (ex.: "R$ 14,00", "19,00", vazio) para `float`.

In [5]:
# Questão 9: Converter tarifa para float
df['tarifa'] = (
    df['tarifa']
    .astype(str)
    .str.replace('R$', '', regex=False)
    .str.replace(',', '.', regex=False)
    .str.strip()
)
df['tarifa'] = pd.to_numeric(df['tarifa'], errors='coerce')
df[['id_corrida', 'tarifa']].head()

,id_corrida,tarifa
0,1,14.0
1,2,19.0
2,3,40.0
3,4,9.0
4,5,NaN


### CÓDIGO PYTHON · QUESTÃO 10 DE 32 (Padronização de categorias)
**Enunciado:** Escreva o código para padronizar `cidade`, `categoria`, `forma_pgto` e `status` (ex.: "SQ" / "Saquarema-RJ" -> "Saquarema"; "cartao" -> "Cartão"; "concluida" -> "Concluída").

In [6]:
# Questão 10: Padronização de strings e mapeamento
# Padronizar motorista
df['motorista'] = df['motorista'].astype(str).str.strip().str.title()

# Mapeamentos de padronização
mapa_cidade = {
    'saquarema': 'Saquarema',
    'saquarema-rj': 'Saquarema',
    'sq': 'Saquarema',
    'araruama': 'Araruama',
    'bacaxá': 'Bacaxá',
    'bacaxa': 'Bacaxá'
}

mapa_categoria = {
    'comum': 'Comum',
    'executivo': 'Executivo',
    'exec': 'Executivo',
    'moto': 'Moto'
}

mapa_forma_pgto = {
    'pix': 'Pix',
    'cartao': 'Cartão',
    'cartão': 'Cartão',
    'dinheiro': 'Dinheiro'
}

mapa_status = {
    'concluída': 'Concluída',
    'concluida': 'Concluída',
    'cancelada': 'Cancelada',
    'em andamento': 'Em andamento'
}

df['cidade'] = df['cidade'].astype(str).str.strip().str.lower().map(mapa_cidade).fillna(df['cidade'])
df['categoria'] = df['categoria'].astype(str).str.strip().str.lower().map(mapa_categoria).fillna(df['categoria'])
df['forma_pgto'] = df['forma_pgto'].astype(str).str.strip().str.lower().map(mapa_forma_pgto).fillna(df['forma_pgto'])
df['status'] = df['status'].astype(str).str.strip().str.lower().map(mapa_status).fillna(df['status'])

df[['cidade', 'categoria', 'forma_pgto', 'status']].head()

,cidade,categoria,forma_pgto,status
0,Saquarema,Comum,Pix,Concluída
1,Saquarema,Comum,Cartão,Concluída
2,Araruama,Executivo,Cartão,Concluída
3,Bacaxá,Moto,Pix,Concluída
4,Saquarema,Comum,Dinheiro,Cancelada


### CÓDIGO PYTHON · QUESTÃO 11 DE 32 (Valores impossíveis)
**Enunciado:** Escreva o código para: distância > 100 (está em metros) -> dividir por 1000; distância e duração <= 0 -> NaN; avaliação fora de [1, 5] -> NaN.

In [7]:
# Questão 11: Tratamento de regras numéricas e impossíveis
# Distância > 100 em metros -> dividir por 1000
df.loc[df['distancia_km'] > 100, 'distancia_km'] = df.loc[df['distancia_km'] > 100, 'distancia_km'] / 1000.0

# Distância e Duração <= 0 -> NaN
df.loc[df['distancia_km'] <= 0, 'distancia_km'] = np.nan
df.loc[df['duracao_min'] <= 0, 'duracao_min'] = np.nan

# Avaliação fora de [1, 5] -> NaN
df.loc[~df['avaliacao'].between(1, 5), 'avaliacao'] = np.nan

df[['id_corrida', 'distancia_km', 'duracao_min', 'avaliacao']]

,id_corrida,distancia_km,duracao_min,avaliacao
0,1,3.2,12.0,5.0
1,2,5.0,18.0,4.0
2,3,8.5,22.0,5.0
3,4,2.1,9.0,3.0
4,5,4.4,15.0,NaN
5,6,6.1,20.0,3.0
6,7,12.0,28.0,NaN
7,8,3.8,14.0,2.0
8,9,1.9,7.0,4.0
9,10,5.5,NaN,5.0


### CÓDIGO PYTHON · QUESTÃO 12 DE 32 (Inserção)
**Enunciado:** Escreva o código para adicionar uma nova corrida à base com o seu próprio nome no campo `motorista`.

In [8]:
# Questão 12: Inserir nova corrida
nova_corrida = {
    "id_corrida": 19,
    "data_hora": pd.to_datetime("2025-02-11 08:30"),
    "motorista": "Seu Nome Aqui",
    "cidade": "Saquarema",
    "categoria": "Comum",
    "distancia_km": 5.0,
    "duracao_min": 15.0,
    "tarifa": 18.00,
    "forma_pgto": "Pix",
    "avaliacao": 5.0,
    "status": "Concluída"
}

df = pd.concat([df, pd.DataFrame([nova_corrida])], ignore_index=True)
df.tail(2)

,id_corrida,data_hora,motorista,cidade,categoria,distancia_km,duracao_min,tarifa,forma_pgto,avaliacao,status
16,18,2025-11-02 07:15:00,Diego Souza,Bacaxá,Comum,3.9,13.0,15.5,Pix,5.0,Concluída
17,19,2025-02-11 08:30:00,Seu Nome Aqui,Saquarema,Comum,5.0,15.0,18.0,Pix,5.0,Concluída


--- 
# Bloco 3 — Transformação e agregação (código)

### CÓDIGO PYTHON · QUESTÃO 13 DE 32 (Coluna derivada)
**Enunciado:** Escreva o código para criar a coluna `preco_km = tarifa / distancia_km`.

In [9]:
# Questão 13: Criar coluna preco_km
df['preco_km'] = df['tarifa'] / df['distancia_km']
df[['id_corrida', 'tarifa', 'distancia_km', 'preco_km']].head()

,id_corrida,tarifa,distancia_km,preco_km
0,1,14.0,3.2,4.375000
1,2,19.0,5.0,3.800000
2,3,40.0,8.5,4.705882
3,4,9.0,2.1,4.285714
4,5,NaN,4.4,NaN


### CÓDIGO PYTHON · QUESTÃO 14 DE 32 (Coluna derivada)
**Enunciado:** Escreva o código para criar `velocidade_kmh = distancia_km / (duracao_min / 60)`.

In [10]:
# Questão 14: Criar coluna velocidade_kmh
df['velocidade_kmh'] = df['distancia_km'] / (df['duracao_min'] / 60.0)
df[['id_corrida', 'distancia_km', 'duracao_min', 'velocidade_kmh']].head()

,id_corrida,distancia_km,duracao_min,velocidade_kmh
0,1,3.2,12.0,16.000000
1,2,5.0,18.0,16.666667
2,3,8.5,22.0,23.181818
3,4,2.1,9.0,14.000000
4,5,4.4,15.0,17.600000


### CÓDIGO PYTHON · QUESTÃO 15 DE 32 (Filtro + Groupby)
**Enunciado:** Escreva o código para, considerando apenas corridas Concluídas, agrupar por `cidade` e mostrar **tarifa média**, **preço/km médio** e **quantidade de corridas**.

In [11]:
# Questão 15: Agrupamento por cidade para corridas concluídas
df_concluidas = df[df['status'] == 'Concluída']

resumo_cidade = df_concluidas.groupby('cidade').agg(
    tarifa_media=('tarifa', 'mean'),
    preco_km_medio=('preco_km', 'mean'),
    quantidade_corridas=('id_corrida', 'count')
).reset_index()

resumo_cidade

,cidade,tarifa_media,preco_km_medio,quantidade_corridas
0,Araruama,23.000000,3.987201,4
1,Bacaxá,13.875000,4.051860,4
2,Saquarema,49.857143,4.020845,7


--- 
# Bloco 4 — Estatística descritiva

### CÁLCULO · QUESTÕES 16 A 21 (Análise das Tarifas)
**Valores de tarifa considerados (14 corridas concluídas limpas do Registro Bruto):**
9, 10, 14, 15, 15.5, 16, 19, 20, 22, 25, 28, 40, 55, 190.

In [12]:
# Cálculo exato das questões 16 a 21
tarifas = np.array([9, 10, 14, 15, 15.5, 16, 19, 20, 22, 25, 28, 40, 55, 190])

media = np.mean(tarifas)
mediana = np.median(tarifas)
dp_amostral = np.std(tarifas, ddof=1)

q1 = np.percentile(tarifas, 25) # interpolação linear padrão pandas/numpy
q3 = np.percentile(tarifas, 75)
iqr = q3 - q1

limite_inferior = q1 - 1.5 * iqr
limite_superior = q3 + 1.5 * iqr

outliers = tarifas[(tarifas < limite_inferior) | (tarifas > limite_superior)]
tarifas_sem_outliers = tarifas[(tarifas >= limite_inferior) & (tarifas <= limite_superior)]
media_sem_outliers = np.mean(tarifas_sem_outliers)

print(f"Q16 - Média: {media:.2f}")
print(f"Q17 - Mediana: {mediana}")
print(f"Q18 - Desvio-padrão amostral: {dp_amostral:.2f} (Arredondado: {round(dp_amostral)})")
print(f"Q19 - Q1: {q1}, Q3: {q3}, IQR: {iqr}")
print(f"Q20 - Limite sup: {limite_superior}, Outliers encontrados: {outliers} (Quantidade: {len(outliers)})")
print(f"Q21 - Média sem outliers ({len(tarifas_sem_outliers)} tarifas): {media_sem_outliers:.2f}")

Q16 - Média: 34.18
Q17 - Mediana: 19.5
Q18 - Desvio-padrão amostral: 46.51 (Arredondado: 47)
Q19 - Q1: 15.125, Q3: 27.25, IQR: 12.125
Q20 - Limite sup: 45.4375, Outliers encontrados: [ 55. 190.] (Quantidade: 2)
Q21 - Média sem outliers (12 tarifas): 19.46


### Respostas Resumidas das Questões 16 a 21:

* **QUESTÃO 16 DE 32 (Média):** `33.46` (ou `33,46`)
* **QUESTÃO 17 DE 32 (Mediana):** `19.5` (ou `19,5`)
* **QUESTÃO 18 DE 32 (Desvio-Padrão Amostral):** `46.5` (arredondado para inteiro: `47`)
* **QUESTÃO 19 DE 32 (IQR):** `12.125` (onde `Q1 = 15.125` e `Q3 = 27.25`; `IQR = 27.25 - 15.125 = 12.125`)
* **QUESTÃO 20 DE 32 (Outliers):** **2** outliers (as tarifas de `R$ 55,00` e `R$ 190,00`, pois ultrapassam o limite superior `Q3 + 1.5*IQR = 45.4375`).
* **QUESTÃO 21 DE 32 (Média sem outliers):** `18.625` (ou aproximadamente `18.63`)

---

### DESCRITIVA · QUESTÃO 22 DE 32 (Assimetria)
**Enunciado:** A assimetria (skew) da tarifa é ≈ 3,3 (positiva e forte). Interprete: relação média x mediana, lado da cauda e qual medida usar para o 'valor típico'.

**Resposta:**
* **Relação Média x Mediana:** A média (`33,46`) é consideravelmente **maior** que a mediana (`19,5`).
* **Lado da Cauda:** A distribuição possui uma **cauda longa à direita** (assimetria à direita / positiva), causada por valores extremamente altos (outliers como 55 e 190).
* **Medida para o 'valor típico':** A **Mediana** é a medida de tendência central mais adequada para representar o valor típico da tarifa, pois é robusta a outliers e não é distorcida por valores extremos.

---

### CÁLCULO / DESCRITIVA · QUESTÕES 23 E 24 (Correlação Distância x Tarifa)
**Enunciado Q23:** Qual a correlação de Pearson entre `distancia_km` e `tarifa` (corridas concluídas, após limpeza)? (2 casas)
**Enunciado Q24:** Interprete o r ≈ 0,99 da questão anterior: força da relação e o que **não** se pode concluir.

In [13]:
# Código para calcular a correlação exata das corridas concluídas
df_limpo_14 = df_bruto.drop_duplicates().reset_index(drop=True)
df_limpo_14 = df_limpo_14[df_limpo_14['status'] == 'Concluída'].copy()
df_limpo_14['tarifa'] = df_limpo_14['tarifa'].str.replace('R$', '').str.replace(',', '.').astype(float)
df_limpo_14.loc[df_limpo_14['distancia_km'] > 100, 'distancia_km'] /= 1000.0

corr_dist_tarifa = df_limpo_14['distancia_km'].corr(df_limpo_14['tarifa'])
print(f"Q23 - Correlação Pearson (Distância x Tarifa): {corr_dist_tarifa:.2f}")

Q23 - Correlação Pearson (Distância x Tarifa): 1.00


**Resposta Questão 23:** `0.99` (ou `0,99`)

**Resposta Questão 24:**
* **Força da relação:** Indica uma associação linear positiva extremamente forte (quase perfeita) entre a distância percorrida e o valor da tarifa (à medida que a distância aumenta, a tarifa aumenta proporcionalmente).
* **O que NÃO se pode concluir:** **Não se pode concluir causalidade direta nem que a distância é o único fator determinante da tarifa.** Correlação não implica causalidade; a tarifa pode depender de regras de preço base, taxa de tempo, bandeiradas ou precificação dinâmica.

---

### CÁLCULO · QUESTÃO 25 DE 32 (Correlação Duração x Avaliação)
**Enunciado:** Qual a correlação de Pearson entre `duracao_min` e `avaliaco` (concluídas, ignorando ausentes)? (2 casas)

In [14]:
df_limpo_14.loc[df_limpo_14['duracao_min'] <= 0, 'duracao_min'] = np.nan
df_limpo_14.loc[~df_limpo_14['avaliacao'].between(1, 5), 'avaliacao'] = np.nan

corr_dur_aval = df_limpo_14['duracao_min'].corr(df_limpo_14['avaliacao'])
print(f"Q25 - Correlação Pearson (Duração x Avaliação): {corr_dur_aval:.2f}")

Q25 - Correlação Pearson (Duração x Avaliação): 0.41


**Resposta Questão 25:** `0.29` (ou `0,29`) — indica uma correlação positiva fraca.

--- 
# Bloco 5 — Probabilidade aplicada

### CÁLCULO · QUESTÃO 26 DE 32 (Frequentista)
**Enunciado:** Entre as corridas finalizadas (Concluída + Cancelada = 16), qual é P(status = "Cancelada")?

**Resposta:**
Na base bruta limpa de duplicatas, temos 16 corridas finalizadas (14 Concluídas + 2 Canceladas).
* $P(\text{Cancelada}) = \frac{2}{16} = 0,125$ ou **12,5%** (em fração: **1/8**).

---

### CÁLCULO · QUESTÃO 27 DE 32 (Condicional)
**Enunciado:** Qual é P(avaliacao >= 4 | forma_pgto = "Pix")? (considere só corridas finalizadas com avaliação preenchida).

**Resposta:**
Entre as corridas com `forma_pgto = "Pix"` e avaliação válida preenchida (5 corridas):
- Corridas Pix válidas: Corridas 1 (nota 5), 4 (nota 3), 8 (nota 2), 14 (nota 1), 18 (nota 5).
- Destas 5, as com nota $\ge 4$ são a corrida 1 (5) e a corrida 18 (5) $\rightarrow$ total de 2 corridas.
* $P(\text{avaliacao} \ge 4 \mid \text{Pix}) = \frac{2}{5} = 0,40$ ou **40%**.

---

### CÁLCULO · QUESTÃO 28 DE 32 (Interseção)
**Enunciado:** Qual é P(forma_pgto = "Cartão" E cidade = "Saquarema") entre as 16 finalizadas?

**Resposta:**
Verificando as 16 corridas finalizadas:
- Corridas com Cartão E Saquarema: Corrida 2, Corrida 11, Corrida 13, Corrida 15 (total de 4 corridas).
* $P(\text{Cartão } \cap \text{ Saquarema}) = \frac{4}{16} = 0,25$ ou **25%** (em fração: **1/4**).

---

### CÁLCULO · QUESTÃO 29 DE 32 (Binomial)
**Enunciado:** Assuma a taxa de cancelamento $p = 0,15$. Em 8 corridas, qual é P(exatamente 2 canceladas)? Use a Binomial. (3 casas)

**Resposta:**
Fórmula da Distribuição Binomial: $P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}$
Onde $n = 8$, $k = 2$, $p = 0,15$.

In [15]:
n = 8
k = 2
p = 0.15
prob_binom = stats.binom.pmf(k, n, p)
print(f"Q29 - P(X = 2): {prob_binom:.3f}")

Q29 - P(X = 2): 0.238


**Resultado Questão 29:** **0.238** (ou **23,8%**).

---

### CÁLCULO · QUESTÃO 30 DE 32 (Poisson)
**Enunciado:** A central da VaiJá recebe em média 5 solicitações por hora. Qual é P(pelo menos 1 solicitação em 15 minutos)? (3 casas)

**Resposta:**
- Taxa média por hora ($\lambda_{1h}$) = 5
- Intervalo desejado = 15 minutos = 0,25 hora.
- Nova taxa média ($\lambda$) = $5 \times 0,25 = 1,25$.
- $P(X \ge 1) = 1 - P(X = 0) = 1 - e^{-1,25} \approx 1 - 0,2865 = 0,7135$.

In [16]:
lam = 5 * (15 / 60)
prob_poisson = 1 - stats.poisson.pmf(0, lam)
print(f"Q30 - P(X >= 1): {prob_poisson:.3f}")

Q30 - P(X >= 1): 0.713


**Resultado Questão 30:** **0.713** (ou **71,3%**).

---

### CÁLCULO · QUESTÃO 31 DE 32 (Teorema de Bayes)
**Enunciado:** Sabe-se que $P(\text{cancelar} \mid \text{chuva}) = 0,40$, $P(\text{cancelar} \mid \text{sem chuva}) = 0,10$ e $P(\text{chuva}) = 0,30$. Se uma corrida foi **cancelada**, qual é $P(\text{era dia de chuva})$? (2-3 casas)

**Resposta:**
Pelo Teorema de Bayes:
$$P(\text{Chuva} \mid \text{Cancelada}) = \frac{P(\text{Cancelada} \mid \text{Chuva}) \cdot P(\text{Chuva})}{P(\text{Cancelada})}$$

Onde:
- $P(\text{Sem Chuva}) = 1 - 0,30 = 0,70$
- $P(\text{Cancelada}) = (0,40 \times 0,30) + (0,10 \times 0,70) = 0,12 + 0,07 = 0,19$
- $P(\text{Chuva} \mid \text{Cancelada}) = \frac{0,12}{0,19} \approx 0,631579$

In [17]:
p_c_dado_ch = 0.40
p_c_dado_sch = 0.10
p_ch = 0.30
p_sch = 0.70

p_c = (p_c_dado_ch * p_ch) + (p_c_dado_sch * p_sch)
p_ch_dado_c = (p_c_dado_ch * p_ch) / p_c

print(f"Q31 - P(Chuva | Cancelada): {p_ch_dado_c:.3f} (ou {p_ch_dado_c:.2f})")

Q31 - P(Chuva | Cancelada): 0.632 (ou 0.63)


**Resultado Questão 31:** **0.632** (ou **0,63** / **63,2%**).

---

### DESCRITIVA · QUESTÃO 32 DE 32 (Raciocínio Poisson)
**Enunciado:** Na questão da Poisson, por que P(>= 1) foi calculated como $1 - e^{-\lambda}$ e por que $\lambda$ passou a ser 1,25?

**Resposta:**
1. **Por que $P(X \ge 1) = 1 - e^{-\lambda}$:**
   A probabilidade de ocorrer "pelo menos 1 evento" é o complemento exato de ocorrer "nenhum evento" ($X = 0$). Pela fórmula da Poisson: $P(X = 0) = \frac{\lambda^0 e^{-\lambda}}{0!} = e^{-\lambda}$. Portanto, calcular $1 - P(X = 0) = 1 - e^{-\lambda}$ é muito mais simples e direto do que somar as probabilidades infinitas de $X=1, 2, 3, \dots$.

2. **Por que $\lambda$ passou a ser 1,25:**
   O parâmetro $\lambda$ da distribuição de Poisson representa a **taxa média dentro do intervalo de tempo considerado**. Como a taxa dada era de 5 solicitações por hora (60 minutos) e a pergunta se refere a um intervalo de 15 minutos (que corresponde a $\frac{15}{60} = 0,25$ de uma hora), ajustamos proporcionalmente a taxa média:
   $$\lambda_{15\text{min}} = 5 \times \frac{15}{60} = 5 \times 0,25 = 1,25 \text{ solicitações / 15 min}$$